### PR TITLE
Changed item-permission message to be per item rather than one global list

### PR DESCRIPTION
**Pull request by xH3LLRAIZ3Rx**
## Description
I decided that it would be better all around to be able to change the permission message per item rather than one global message.

## Changes
Added permission-message string list
Removed tooltips.item-message in messages class

## Testability
- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.

### CODE REVIEW

**Comment by ajan-12** [comment]
Told ya lol

**Comment by xH3LLRAIZ3Rx** [comment]
> Told ya lol

Yup lol, I ended up needing it.

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java [hunk 1]
```diff
@@ -50,6 +50,7 @@
 	private boolean replacing = false;
 	private boolean addon = false;
 	private String permission = "";
+	private String[] permissionMessage = {"&4&lLOCKED", "", "&rYou do not have Permission", "&rto access this Item"};
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: Should rather be "noPermissionTooltip" as permissionMessage is misleading...

File: src/me/mrCookieSlime/Slimefun/SlimefunGuide.java [hunk 2]
```diff
@@ -713,8 +713,8 @@ public void run(final Player p) {
 							index++;
 						}
 						else {
-							List<String> tooltip = Messages.local.getTranslation("tooltips.item-permission");
-						    menu.addItem(index, new CustomItem(Material.BARRIER, StringUtils.formatItemName(sfitem.getItem(), false), tooltip.toArray(new String[tooltip.size()])));
+							List<String> message = SlimefunPlugin.getItemCfg().getStringList(sfitem.getID() + ".permission-message");
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/SlimefunGuide.java]: Why would you read from the config if you implemented a getter in SlimefunItem...??

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java (hunk 1 above)
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: This value is read-only atm, keeping it in RAM for all the thousands of Slimefunitems makes zero sense in your design, you should write to it in the register() block were all the other values are set as well.

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java [hunk 3]
```diff
@@ -178,6 +179,7 @@ public SlimefunItem(Category category, ItemStack item, String id, RecipeType rec
 	 * @since 4.1.11
 	 */
 	public String getPermission() 			{		return permission;		}
+	public String[] getPermissionMessage()    {       return permissionMessage;       }
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: This getter makes no sense if it just returns the same constant default value for each object.
You should read the value from config in .register()

Comment (xH3LLRAIZ3Rx) [reply src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: Honestly I am not sure why I put a getter

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java (hunk 1 above)
Comment (xH3LLRAIZ3Rx) [reply src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: Would this work?

File: src/me/mrCookieSlime/Slimefun/SlimefunGuide.java (hunk 2 above)
Comment (xH3LLRAIZ3Rx) [reply src/me/mrCookieSlime/Slimefun/SlimefunGuide.java]: What about now?

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java (hunk 1 above)
Comment (xH3LLRAIZ3Rx) [reply src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: Alright

**Comment by TheBusyBiscuit** [comment]
Not going to submit another review, I already said what should be changed.
You actually did the opposite XD

I said that your getter and field are pointless if you don't write to them.
And that you shouldn't read from the config hundreds of times when the guide is opened...

Instead USE the field and getter how they should be used, not remove them...

**Comment by xH3LLRAIZ3Rx** [comment]
> Not going to submit another review, I already said what should be changed.
> You actually did the opposite XD
> 
> I said that your getter and field are pointless if you don't write to them.
> And that you shouldn't read from the config hundreds of times when the guide is opened...
> 
> Instead USE the field and getter how they should be used, not remove them...

Sorry dude, I have not slept yet so I am not as focused as I should be, some caffeine will fix it though.

**Comment by ajan-12** [comment]
Didn't check your code fully but this is how I'd do it:
Set default value on register.
Read values on postRegister.
When item is used check if player has unlocked the item and if not send the message read in postRegister.
(Messages wouldn't be tooltip but instead sent to player via chat.)
Note: Changing tooltips may cause issues in the future.

**Comment by TheBusyBiscuit** [comment]
@ajan-12 If you use postRegister than every implementation in Slimefun that overrides that needs to get a super.postRegister() call.
I agree, this just has to be considered and done too then.

**Comment by xH3LLRAIZ3Rx** [comment]
> 
> 
> @ajan-12 If you use postRegister than every implementation in Slimefun that overrides that needs to get a super.postRegister() call.
> I agree, this just has to be considered and done too then.

Honestly the current code works, so I think you should merge it then let him or someone else do the rest because I don't know enough java to do the rest.

File: src/me/mrCookieSlime/Slimefun/SlimefunGuide.java [hunk 4]
```diff
@@ -713,8 +713,8 @@ public void run(final Player p) {
 							index++;
 						}
 						else {
-							List<String> tooltip = Messages.local.getTranslation("tooltips.item-permission");
-						    menu.addItem(index, new CustomItem(Material.BARRIER, StringUtils.formatItemName(sfitem.getItem(), false), tooltip.toArray(new String[tooltip.size()])));
+							List<String> message = SlimefunItem.getNoPermissionTooltip(sfitem.getItem());
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/SlimefunGuide.java]: Oh god no.
Why would you loop through all Slimefun items to find the Slimefun Item that you already had...???

Just do sfItem.getNoPermissionTooltip() ...

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java [hunk 5]
```diff
@@ -336,6 +340,15 @@ public static SlimefunItem getByItem(ItemStack item) {
 		return null;
 	}
 
+	public static List<String> getNoPermissionTooltip(ItemStack item) {
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: Remove this method.

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java [hunk 6]
```diff
@@ -243,6 +246,7 @@ public void register(boolean slimefun) {
 				this.enchantable = SlimefunPlugin.getItemCfg().getBoolean(this.id + ".allow-enchanting");
 				this.disenchantable = SlimefunPlugin.getItemCfg().getBoolean(this.id + ".allow-disenchanting");
 				this.permission = SlimefunPlugin.getItemCfg().getString(this.id + ".required-permission");
+				this.noPermissionTooltip = SlimefunPlugin.getItemCfg().getStringList(this.id + ".permission-message");
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: The path should match the field name "no-permission-tooltip"

File: src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java [hunk 7]
```diff
@@ -216,7 +218,8 @@ public void register(boolean slimefun) {
 			SlimefunPlugin.getItemCfg().setDefaultValue(this.id + ".allow-enchanting", this.enchantable);
 			SlimefunPlugin.getItemCfg().setDefaultValue(this.id + ".allow-disenchanting", this.disenchantable);
 			SlimefunPlugin.getItemCfg().setDefaultValue(this.id + ".required-permission", this.permission);
-			
+			SlimefunPlugin.getItemCfg().setDefaultValue(this.id + ".permission-message", new String[] {"&4&lLOCKED", "", "&rYou do not have Permission", "&rto access this Item"});
```
Comment (TheBusyBiscuit) [inline src/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/SlimefunItem.java]: The path should match the field name "no-permission-tooltip"

File: src/me/mrCookieSlime/Slimefun/SlimefunGuide.java (hunk 4 above)
Comment (xH3LLRAIZ3Rx) [reply src/me/mrCookieSlime/Slimefun/SlimefunGuide.java]: Yeah I was tired when I did that, fixed ;)

**Pull request merged**